extensions/SampleExtension/SettingsProvider/SettingsUIController.cs

```diff
@@ -0,0 +1,122 @@
+    {
+        Log.Debug($"Dispose");
+        _settingsUI?.Update(null, null, null);
+    }
+
+    public ProviderOperationResult Initialize(IExtensionAdaptiveCard extensionUI)
+    {
+        Log.Debug($"Initialize");
+        _settingsUI = extensionUI;
+        return UpdateCard();
+    }
+
+    public IAsyncOperation<ProviderOperationResult> OnAction(string action, string inputs)
+    {
+#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+        return Task.Run(async () =>
+        {
+            try
+            {
+                Log.Information($"OnAction() called with {action}");
+                Log.Debug($"inputs: {inputs}");
+                var actionObject = JsonNode.Parse(action);
+                var verb = actionObject?["verb"]?.GetValue<string>() ?? string.Empty;
+                Log.Debug($"Verb: {verb}");
+                switch (verb)
+                {
+                    case "OpenLogs":
+                        FileHelper.OpenLogsLocation();
+                        break;
+
+                    default:
+                        Log.Warning($"Unknown verb: {verb}");
+                        break;
+                }
+
+                return UpdateCard();
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Unexpected failure handling settings action.");
+                return new ProviderOperationResult(ProviderOperationStatus.Failure, ex, ex.Message, ex.Message);
+            }
+        }).AsAsyncOperation();
+#pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
+    }
+
+    private ProviderOperationResult UpdateCard()
+    {
+        try
+        {
+            var settingsCardData = new SettingsCardData
+            {
+            };
+
+            return _settingsUI!.Update(
+                GetTemplate(),
+                JsonSerializer.Serialize(settingsCardData, SettingsCardSerializerContext.Default.SettingsCardData),
+                "SettingsPage");
+        }
+        catch (Exception ex)
+        {
+            Log.Error(ex, "Failed to update settings card");
+            return new ProviderOperationResult(ProviderOperationStatus.Failure, ex, ex.Message, ex.Message);
+        }
+    }
+
+    private string GetTemplate()
+    {
+        if (_template is not null)
+        {
+            Log.Debug("Using cached template.");
+            return _template;
+        }
+
+        try
+        {
+            var path = Path.Combine(AppContext.BaseDirectory, @"SettingsProvider\SettingsCardTemplate.json");
+            var template = File.ReadAllText(path, Encoding.Default) ?? throw new FileNotFoundException(path);
+            template = Resources.ReplaceIdentifiers(template, GetSettingsCardResourceIdentifiers(), Log);
+            Log.Debug($"Caching template");
+            _template = template;
+            return _template;
+        }
+        catch (Exception e)
```

Comment:
Such a nit, but something about this being a sample makes me want to be hyper consistent -- other exceptions in this file use `ex` rather than `e`.